### ISSUE TITLE
Constant pattern vs governing type that is unconstrained type parameter

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#24550]
The following in currently rejected:

``` c#
    void M<T>(T t)
    {
        switch (t)
        {
            // Cannot implicitly converty type 'int' to 'T'
            case 3: break;
            // Cannot convert null to type parameter 'T' because it could be a non-nullable value type. Consider using 'default(T)' instead.
            case null: break;
        }
        // Cannot implicitly converty type 'int' to 'T'
        if (t is 3) { }
        // Cannot convert null to type parameter 'T' because it could be a non-nullable value type. Consider using 'default(T)' instead.
        if (t is null) { }
    }
```

However, these are perfectly reasonable pattern-matching operations. The spec and implementation should change to permit this.

**Comment by gafter** [comment]
This will be addressed in C# 8.0 with the change https://github.com/dotnet/csharplang/issues/1284

**Comment by gafter** [comment]
Fixed in https://github.com/dotnet/roslyn/pull/25995
